### PR TITLE
Raw alpha

### PR DESCRIPTION
**Pull request by dmbavykin**

### CODE REVIEW

File: lib/codebreaker/console.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+require_relative 'statistic'
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb]: Add an empty line after magic comments.

File: lib/codebreaker/console.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb]: Extra blank line detected.

File: lib/codebreaker/console.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+class Console
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb]: Missing top-level class documentation comment.

File: lib/codebreaker/console.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  include Statistic
+
+  PHRASES_PATH = 'lib/codebreaker/src/phrases.yml'.freeze
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb]: Do not freeze immutable objects, as freezing them has no effect.

File: lib/codebreaker/console.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    output(:rules)
+    difficulty_rules
+  end
+
+  def output_statistics
+    handle_statistics_for_output.each { |record| output(record) }
+  end
+
+  def round_question(attempts)
+    ask("#{@phrases[:round_question]}#{attempts.to_s}")
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb]: Redundant use of `Object#to_s` in interpolation.

File: lib/codebreaker/console.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  def loose(secret_code)
+    output(@phrases[:loose] << secret_code.join)
+  end
+
+  private
+
+  def difficulty_rules
+    Game::DIFFICULTIES.each do |diff_name, params|
+      output("#{diff_name}: #{params[:attempts]} attempts and #{params[:hints]} hints")
```
Comment (DontBeSoSure) [inline lib/codebreaker/console.rb:39]: Line is too long. [87/80]

File: lib/codebreaker/game.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+class Game
```
Comment (DontBeSoSure) [inline lib/codebreaker/game.rb:3]: - Class has too many lines. [105/100]
- Missing top-level class documentation comment.

File: lib/codebreaker/game.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
```
Comment (DontBeSoSure) [inline lib/codebreaker/game.rb]: Extra empty line detected at class body beginning.

File: lib/codebreaker/game.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  def equal_codes?
+    secret_code.join == @user_answer
+  end
+
+  def attempts
+    @attempts ||= DIFFICULTIES.dig(@difficulty_name, :attempts)
+  end
+
+  def hints
+    @hints ||= secret_code.shuffle.take(DIFFICULTIES.dig(@difficulty_name, :hints))
```
Comment (DontBeSoSure) [inline lib/codebreaker/game.rb:102]: Line is too long. [83/80]

File: lib/codebreaker/game.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+  def win
+    console.output(:win)
+    save_score if console.dichotomy_question?(:save_score)
+    console.dichotomy_question?(:new_game) ? Game.new : console.output(:goodbye)
+  end
+
+  def save_score
+    name = console.ask(:username)
+    score_data = { name: name, difficulty: @difficulty_name, attempts: attempts, hints: hints.size }
```
Comment (DontBeSoSure) [inline lib/codebreaker/game.rb]: Line is too long. [100/80]

File: lib/codebreaker/statistic.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+module Statistic
```
Comment (DontBeSoSure) [inline lib/codebreaker/statistic.rb]: - Missing top-level module documentation comment.
- Missing magic comment `# frozen_string_literal: true`.

File: lib/codebreaker/statistic.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
```
Comment (DontBeSoSure) [inline lib/codebreaker/statistic.rb]: Extra empty line detected at module body beginning.

File: lib/codebreaker/statistic.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  STATISTICS_PATH = 'lib/codebreaker/src/statistics.yml'.freeze
+
+  def statistics
+    @stats ||= YAML.load_file(STATISTICS_PATH)
+  end
+
+  def handle_statistics_for_output
+    statistics.map do |record|
+      "#{record[:name]} won the game on #{record[:difficulty]} level and still had #{record[:attempts]} attempts and #{record[:hints]} hints"
```
Comment (DontBeSoSure) [inline lib/codebreaker/statistic.rb]: Line is too long. [141/80]

File: lib/codebreaker/statistic.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    end
+  end
+
+  def save_score(data)
+    File.new(STATISTICS_PATH, 'w+') unless File.exist?(STATISTICS_PATH)
+    record = statistics || []
+    File.write(STATISTICS_PATH, (record << data).to_yaml)
+  end
+
```
Comment (DontBeSoSure) [inline lib/codebreaker/statistic.rb]: Extra empty line detected at module body end.

File: spec/codebreaker_spec.rb [hunk 15]
```diff
@@ -1,9 +1,153 @@
-RSpec.describe Codebreaker do
-  it "has a version number" do
-    expect(Codebreaker::VERSION).not_to be nil
+# frozen_string_literal: true
+
+RSpec.describe Game do
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb:3]: Block has too many lines. [122/25]

File: spec/codebreaker_spec.rb [hunk 16]
```diff
@@ -1,9 +1,153 @@
...
+      subject.instance_variable_set(:@hints_count, 2)
+    end
+
+    it 'return last number from hints array' do
+      expected_hint = subject.instance_variable_get(:@hints).last
+      expect(subject.send(:take_a_hint!)).to eq(expected_hint)
+    end
+
+    it 'take one number as hint and remove it from array' do
+      expect { subject.send(:take_a_hint!) }.to change { subject.instance_variable_get(:@hints).size }.by(-1)
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb:49]: Line is too long. [109/80]

File: spec/codebreaker_spec.rb [hunk 17]
```diff
@@ -1,9 +1,153 @@
...
+    end
+
+    it 'decrement hints count by one' do
+      expect { subject.send(:take_a_hint!) }.to change { subject.instance_variable_get(:@hints_count) }.by(-1)
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Line is too long. [110/80]

File: spec/codebreaker_spec.rb [hunk 18]
```diff
@@ -1,9 +1,153 @@
...
+    it 'handles correct number' do
+      subject.instance_variable_set(:@user_answer, '1234')
+      expect(subject).to receive(:handle_code)
+      subject.send(:handle_answer)
+    end
+  end
+
+  context '#validate_answer' do
+    it 'tells about input error' do
+      expect { subject.send(:validate_answer) }.to output("Put 4-digital number\n").to_stdout
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Line is too long. [93/80]

File: spec/codebreaker_spec.rb [hunk 19]
```diff
@@ -1,9 +1,153 @@
...
+      subject.instance_variable_set(:@user_answer, '1234')
+      expect(subject.send(:validate_answer)).to be true
+    end
+  end
+
+  context '#setup_the_difficulty' do
+    it 'setup easy level of difficulty' do
+      subject.instance_variable_set(:@difficulty, Game::DIFFICULTY[:easy])
+      subject.send(:setup_the_difficulty)
+      expect(subject.instance_variable_get(:@attempts_count)).to eq(Game::DIFFICULTY.dig(:easy, :attempts))
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Line is too long. [107/80]

File: spec/codebreaker_spec.rb [hunk 20]
```diff
@@ -1,9 +1,153 @@
...
+      expect(subject.instance_variable_get(:@hints_count)).to eq(Game::DIFFICULTY.dig(:easy, :hints))
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Line is too long. [101/80]

File: spec/codebreaker_spec.rb [hunk 21]
```diff
@@ -1,9 +1,153 @@
...
+    it 'setup hard level of difficulty' do
+      subject.instance_variable_set(:@difficulty, Game::DIFFICULTY[:hard])
+      subject.send(:setup_the_difficulty)
+      expect(subject.instance_variable_get(:@attempts_count)).to eq(10)
+      expect(subject.instance_variable_get(:@hints_count)).to eq(1)
+    end
+  end
+
+  context '#difficulty' do
+    it "tells about spelling error and calls itself again" do
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/codebreaker_spec.rb [hunk 22]
```diff
@@ -1,9 +1,153 @@
...
+      allow(subject).to receive(:choose_the_difficulty)
+      expect(subject).to receive(:choose_the_difficulty)
+      subject.send(:handle_difficulty, 'test')
+    end
+
+    it "calls setuping method with easy lvl" do
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/codebreaker_spec.rb [hunk 23]
```diff
@@ -1,9 +1,153 @@
...
+      allow(subject).to receive(:setup_the_difficulty)
+      expect(subject).to receive(:setup_the_difficulty)
+      subject.send(:handle_difficulty, 'easy')
+    end
+
+    it "calls setuping method with medium lvl" do
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/codebreaker_spec.rb [hunk 24]
```diff
@@ -1,9 +1,153 @@
...
+      allow(subject).to receive(:setup_the_difficulty)
+      expect(subject).to receive(:setup_the_difficulty)
+      subject.send(:handle_difficulty, 'medium')
+    end
+
+    it "calls setuping method with hard lvl" do
```
Comment (DontBeSoSure) [inline spec/codebreaker_spec.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/codebreaker_spec.rb [hunk 25]
```diff
@@ -1,9 +1,153 @@
-RSpec.describe Codebreaker do
-  it "has a version number" do
-    expect(Codebreaker::VERSION).not_to be nil
+# frozen_string_literal: true
+
+RSpec.describe Game do
+  context '#make_secret_code' do
+    before do
```
Comment (hf6cTXNPfyPgOBK6) [inline spec/codebreaker_spec.rb]: `let(:secret_code) { subject.send(:make_secret_code) } `

**Review by hf6cTXNPfyPgOBK6 — CHANGES_REQUESTED** [review]

**Pull request merged**